AsyncDisplayKit/Details/_ASDisplayView.mm

```diff
@@ -375,3 +376,81 @@ - (UIView *)preferredFocusedView
 }
 #endif
 @end
+
+
+#pragma mark - Accessibility
+
+static const char *ASDisplayNodeAssociatedNodeKey = "ASAssociatedNode";
+
+@implementation UIAccessibilityElement (_ASDisplayView)
+
+- (void)setAsyncdisplaykit_node:(ASDisplayNode *)node
+{
+  objc_setAssociatedObject(self, ASDisplayNodeAssociatedNodeKey, node, OBJC_ASSOCIATION_ASSIGN); // Weak reference to avoid cycle, since the node retains the layer.
+  
+  // Update UIAccessibilityElement properties from node
+  self.accessibilityLabel = node.accessibilityLabel;
+  self.accessibilityHint = node.accessibilityHint;
+  self.accessibilityValue = node.accessibilityValue;
+  self.accessibilityTraits = node.accessibilityTraits;
+}
+
+- (ASDisplayNode *)asyncdisplaykit_node
+{
+  return objc_getAssociatedObject(self, ASDisplayNodeAssociatedNodeKey);
+}
+
+@end
+
+@implementation _ASDisplayView (UIAccessibilityContainer)
+
+#pragma mark - UIAccessibility
+
+- (NSArray *)accessibleElements
+{
+  ASDisplayNode *asyncDisplayKitNode = self.asyncdisplaykit_node;
+  if ( _accessibleElements != nil ) {
+    return _accessibleElements;
+  }
+  
+  _accessibleElements = [[NSMutableArray alloc] init];
+  
+  // Create UI accessiblity elements for each subnode that represent the subnode within the accessibility container
+  for (ASDisplayNode *subnode in asyncDisplayKitNode.subnodes) {
+    if (subnode.isAccessibilityElement || [subnode accessibilityElementCount] > 0) {
+      UIAccessibilityElement *accessibilityElement = [[UIAccessibilityElement alloc] initWithAccessibilityContainer:self];
+      accessibilityElement.asyncdisplaykit_node = subnode;
+      [_accessibleElements addObject:accessibilityElement];
+    }
+  }
+  
+  return _accessibleElements;
+}
+
+- (NSInteger)accessibilityElementCount
+{
+  return [self accessibleElements].count;
```

Comment:
Even though it would be quite suboptimal for the first implementation to regenerate the array on each call to this, I suspect that it will not cause significant performance problems especially because Voice over itself adds fairly significant complexity and does not allow Long scrolling animations where frame drops are essential. There is also essentially no notion of touch latency with voiceovers activation mechanism.

In a future upgrade, it seems to me that the correct way to optimize this is to use the the environment system, doing some form of upward propagation of the count of accessibility elements at each level.  Changing a property such as is accessibility element may trigger a new propagation to update the state of the tree, interestingly this may be a use case where both upward and downward propagation is required in certain situations.

I don't think we need to schedule this optimization work to occur in the next few months however, and it would be interesting to do some profiling on the voiceover mode just to check where the bottlenecks are anyway, as there could be other parts of the framework about are even more useful to optimize for this mode